CrystalMatch/dls_focusstack/focus/fourier.py

```diff
@@ -11,27 +11,34 @@ def __init__(self, input_array):
         self.array = input_array
 
     def runFFT(self):
-        return self.fourier()
+        return self.fourier_bandpass()
 
-    # algorithm provided by Charles - a slight modification of the optimal sizes
-    # has been added to speed up the procedure
-    def fourier(self):
-        """Function used to calculate FFT of the image - created by Charles."""
+    # algorithm provided by Charles
+    def fourier_bandpass(self):
+        """Function used to calcualte FFT of the image and apply a bandpass filter"""
         rows, cols = self.array.shape
         nrows = cv2.getOptimalDFTSize(rows)
         ncols = cv2.getOptimalDFTSize(cols)
         nimg = np.zeros((nrows, ncols))
         nimg[:rows, :cols] = self.array
 
-        data_fft1 = np.fft.rfft2(nimg) #the opcv version uses fft2 not rfft2
-        fft_abs1 = np.abs(data_fft1).copy()
-        h1, w1 = fft_abs1.shape
-        # -1 is the last element, last excluded
-        # zero frequency is in the top left corner
-        part1 = fft_abs1[int(0.05 * h1) : int(0.95 * h1)-1, int(0.05 * w1) : -1]
-        output = np.mean(part1)
-
-        return output
-
-
-
+        data_fft = np.fft.rfft2(nimg)
+        fft_abs = np.abs(data_fft)
+
+        h, w = fft_abs.shape
+        l = int(min(h/2, w)) # rfft2 returns half the width of fft2
+        min_l = 0.2 * l
+        max_l = 0.6 * l
+
+        def filter_indices(x, y, max_x, max_y):
+            x2, y2 = x*x, y*y
+            return (x > 0.05 * max_x) \
+                    & (y > 0.05 * max_y) \
+                    & (min_l * min_l < x2 + y2) \
+                    & (x2 + y2 < max_l * max_l)
+
```

Comment:
This is where the 'ring' is defined. Frequencies within the ring are going to be kept the rest is discarded.